astroquery/heasarc/tests/parametrization.py

```diff
@@ -0,0 +1,109 @@
+import os
+import glob
+import hashlib
+import requests
+import pytest
+from ... import log
+
+"""
+This is an attempt to reduce code duplication between remote and local tests.
+
+if there is test data:
+    runs as usual, except all tests have two versions with the same code: remote and local
+else
+    runs remote test patched so that the test data is stored in a temporary directory.
+    advice is given to copy the newly generated test data into the repository
+"""
+
+
+class MockResponse:
+    def __init__(self, text):
+        self.text = text
+        self.content = text.encode()
+
+
+def data_path(filename, output=False):
+    if output:
+        data_dir = os.path.join(
+            os.getenv("TMPDIR", "/tmp"), "astroquery-heasarc-saved-data"
+        )
+        os.makedirs(data_dir, exist_ok=True)
+    else:
+        data_dir = os.path.join(os.path.dirname(__file__), "data")
+
+    return os.path.join(data_dir, filename + ".dat")
+
+
+def filename_for_request(url, params, output=False):
+    fileid = hashlib.md5(str((url, sorted(params.items()))).encode()).hexdigest()[:8]
+    return data_path(fileid, output=output)
+
+
+# TODO: are get_mockreturn args up-to-date in example in https://astroquery.readthedocs.io/en/latest/testing.html ?
+def get_mockreturn(session, method, url, params=None, timeout=10, **kwargs):
+
+    filename = filename_for_request(url, params)
+    try:
+        content = open(filename, "rt").read()
+    except FileNotFoundError:
+        log.error(
+            f'no stored mock data in {filename} for url="{url}" and params="{params}"'
+            "perhaps you need to clean test data and regenerate it? "
+            "It will be regenerated automatically if cleaned, try `rm -fv astroquery/heasarc/tests/data/* build`"
+        )
+        raise
+
+    return MockResponse(content)
+
+
+def save_response_of_get(session, method, url, params=None, timeout=10, **kwargs):
+
+    text = requests.Session._original_request(
+        session, method, url, params=params, timeout=timeout
+    ).text
+
+    filename = filename_for_request(url, params, output=True)
+
+    with open(filename, "wt") as f:
+        log.info(f'saving output to {filename} for url="{url}" and params="{params}"')
+        log.warning(
+            f"you may want to run `cp -fv {os.path.dirname(filename)}/* astroquery/heasarc/tests/data/; rm -rfv build`"
```

Comment:
This suggestive warning is a good idea, but we should also have it point to some narrative documentation explaining why

Reply:
also might want to remind us that if we're `cp`ing into the source tree, we want to `git add` the changes.

Reply:
> This suggestive warning is a good idea, but we should also have it point to some narrative documentation explaining why

Ok, but documentation where?

Reply:
https://astroquery.readthedocs.io/en/latest/testing.html

Reply:
> also might want to remind us that if we're `cp`ing into the source tree, we want to `git add` the changes.

Ok, added.

Reply:
> https://astroquery.readthedocs.io/en/latest/testing.html

I was thinking about something like that, but then this is still in the heasarc module? Does it make sense to add documentation on module-specific (for now) code in such a generic place?
If I add it there, maybe then better to follow the other option you suggested, and put it to `utils` in this PR?
Or add the extra details and documentation in the other PR.
As you wish.

Reply:
ah, no, you're quite right - we can delay adding to the narrative docs until that future PR

Reply:
Ok, noted.